src/lib/comms/sol-socket-impl-zephyr.c

```diff
@@ -0,0 +1,438 @@
+    }
+    nano_task_sem_give(&socket->lock);
+
+    if (!buf)
+        return buflen;
+
+    if (buflen > len)
+        buflen = len;
+
+    if (cliaddr) {
+        cliaddr->family = SOL_NETWORK_FAMILY_INET6;
+        cliaddr->port = uip_ntohs(NET_BUF_UDP(netbuf)->srcport);
+        memcpy(cliaddr->addr.in6, &NET_BUF_IP(netbuf)->srcipaddr, sizeof(cliaddr->addr.in6));
+    }
+
+    memcpy(buf, ip_buf_appdata(netbuf), buflen);
+
+    ip_buf_unref(netbuf);
+
+    return buflen;
+}
+
+/* HACK HACK HACK HACK
+ * We should not need to include this header anywhere around here, but in order
+ * to make net_reply() work for our needs, the buf we pass to it must contain
+ * a valid udp_conn set, and to get that one, we need to peek into the
+ * declaration of simple_udp_connection.
+ */
+#include <ip/simple-udp.h>
+
+static int
+sol_socket_zephyr_sendmsg(struct sol_socket *s, const void *buf, size_t len, const struct sol_network_link_addr *cliaddr)
+{
+    struct sol_socket_zephyr *socket = (struct sol_socket_zephyr *)s;
+    struct net_context *ctx;
+    struct net_tuple *tuple;
+    struct net_buf *netbuf;
+    uint8_t *ptr;
+
+    ctx = socket->unicast_context;
+    SOL_NULL_CHECK(ctx, -ENOTCONN);
+
+    tuple = net_context_get_tuple(ctx);
+
+    netbuf = ip_buf_get_tx(ctx);
+    SOL_NULL_CHECK(netbuf, -ENOMEM);
+
+    /* We set address and port on the packet as if we had received this from
+     * the client we want to send to because net_reply() will revert them
+     * before sending */
+    memcpy(&NET_BUF_IP(netbuf)->srcipaddr, cliaddr->addr.in6, sizeof(cliaddr->addr.in6));
+    memcpy(&NET_BUF_IP(netbuf)->destipaddr, &tuple->local_addr->in6_addr, sizeof(tuple->local_addr->in6_addr));
+    NET_BUF_UDP(netbuf)->srcport = uip_htons(cliaddr->port);
+    NET_BUF_UDP(netbuf)->destport = uip_htons(tuple->local_port);
+    uip_set_udp_conn(netbuf) = net_context_get_udp_connection(ctx)->udp_conn;
+
+    ptr = net_buf_add(netbuf, len);
+    memcpy(ptr, buf, len);
+    ip_buf_appdatalen(netbuf) = len;
+
+    if (net_reply(ctx, netbuf) < 0) {
+        ip_buf_unref(netbuf);
+        return -EIO;
+    }
+
+    return len;
+}
+
+static int
+sol_socket_zephyr_join_group(struct sol_socket *s, int ifindex, const struct sol_network_link_addr *group)
+{
+    struct sol_socket_zephyr *socket = (struct sol_socket_zephyr *)s;
+    struct net_context *ctx;
+    struct net_tuple *tuple;
+    struct net_addr bindaddr;
+    int ret;
+
+    if (group->family != SOL_NETWORK_FAMILY_INET6)
+        return -EAFNOSUPPORT;
+
+    tuple = net_context_get_tuple(socket->unicast_context);
+    SOL_NULL_CHECK(tuple, -EINVAL);
+
+    bindaddr.family = sol_network_sol_to_af(group->family);
+    memcpy(&bindaddr.in6_addr, group->addr.in6, sizeof(bindaddr.in6_addr));
+
+    ctx = net_context_get(IPPROTO_UDP, NULL, 0, &bindaddr, tuple->local_port);
+    SOL_NULL_CHECK(ctx, -ENOBUFS);
+
+    ret = sol_ptr_vector_append(&socket->mcast_contexts, ctx);
```

Comment:
On Wed, 30 Mar 2016 07:12:52 -0700, Gustavo Lima Chaves wrote:
> > +    struct net_addr bindaddr;
> > +    int ret;
> > +
> > +    if (group->family != SOL_NETWORK_FAMILY_INET6)
> > +        return -EAFNOSUPPORT;
> > +
> > +    tuple = net_context_get_tuple(socket->unicast_context);
> > +    SOL_NULL_CHECK(tuple, -EINVAL);
> > +
> > +    bindaddr.family = sol_network_sol_to_af(group->family);
> > +    memcpy(&bindaddr.in6_addr, group->addr.in6, sizeof(bindaddr.in6_addr));
> > +
> > +    ctx = net_context_get(IPPROTO_UDP, NULL, 0, &bindaddr, tuple->local_port);
> > +    SOL_NULL_CHECK(ctx, -ENOBUFS);
> > +
> > +    ret = sol_ptr_vector_append(&socket->mcast_contexts, ctx);
> 
> We don't lock on the mcast sockets list too?

Oh for fucks sake... yes, we should, and it's going to be a beautiful
mess.